server/tests/validation.specs/user.spec.js

```diff
@@ -0,0 +1,113 @@
+import mocha from "mocha";
+import chai from "chai";
+import sinon from "sinon";
+import sinonChai from "sinon-chai";
+import user from "../../middlewares/validations/user";
+
+const expect = chai.expect;
+chai.use(sinonChai);
+
+describe("User input validation Middleware", () => {
+  it("should not return an error when all the inputs are valid", done => {
+    const value = {
+      firstname: "mss",
+      lastname: "mss",
+      password: "Ass.2qqqq",
+      email: "d@d.com"
+    };
+    let req = {
+      body: JSON.stringify(value)
+    };
+    let res = {};
+    user.validateUserInput(req, res, () => {
+      done();
+    });
+  });
+
+  it("should return an error when firstname is invalid", done => {
+    const value = {
+      firstname: "",
+      lastname: "mss",
+      password: "Ass.2qqqq",
+      email: "d@d.com"
+    };
+    let req = {
+      body: JSON.stringify(value)
+    };
+    let res = {
+      status: code => {
+        return {
+          json: data => {
+            return data;
+          }
+        };
+      }
+    };
+    let statusSpy = sinon.spy(res, "status");
+
+    user
+      .validateUserInput(req, res, () => {})
+      .then(response => {
+        statusSpy.restore();
+        expect(statusSpy).to.have.been.calledOnce;
+        done();
+      })
+      .catch(error => console.log("error", error));
+  });
+
+  it("should return an error when lastname is invalid", done => {
+    const value = {
+      firstname: "john",
+      lastname: "",
+      password: "Ass.2qqqq",
+      email: "d@d.com"
+    };
+    let req = {
+      body: JSON.stringify(value)
+    };
+    let res = {
+      status: code => {
+        return {
+          json: data => {
+            return data;
+          }
+        };
+      }
+    };
+    let statusSpy = sinon.spy(res, "status");
+
+    user
+      .validateUserInput(req, res, () => {})
+      .then(response => {
+        statusSpy.restore();
+        expect(statusSpy).to.have.been.calledOnce;
+        done();
+      })
+      .catch(error => console.log("error", error));
+  });
+
+  it("should return an error when all the inputs are invalid", done => {
+    const value = {};
+    let req = {
+      body: JSON.stringify(value)
+    };
+    let res = {
+      status: code => {
+        return {
+          json: data => {
+            return data;
+          }
+        };
+      }
+    };
+    let statusSpy = sinon.spy(res, "status");
+
+    user
+      .validateUserInput(req, res, () => {})
+      .then(response => {
+        statusSpy.restore();
+        expect(statusSpy).to.have.been.calledOnce;
```

Comment:
This is not enough testing, the response object has a lot more things we can test